### ISSUE TITLE
Saving Parse.Object with attribute containing array with strings leads to infinite loop

### ISSUE THREAD

**Issue by Gameleon12** [ParsePlatform/Parse-SDK-JS#253]
Currently there is an issue with saving Parse.Object instances to the database using the Javascript SDK.
Any save action for an object containing an array of strings produces a [RangeError: Maximum call stack size exceeded] error.

#### Logs/Trace
Code:
```
var currentArticle= new Article();
currentArticle.set("article_slug",slug); //24 character string
currentArticle.set("article_title",articleTitle);
currentArticle.set("labels",["All","News"]);
try{
  console.log("Start saving object");
  return currentArticle.save();
}
catch(error)
{
  console.log("Catch error block:");
  console.error(error);
}
```
Logs:
```
[[Wed 06 Apr 2016 17:21:24]] [LOG]   Start saving object
[[Wed 06 Apr 2016 17:21:24]] [LOG]   Catch error block:
[[Wed 06 Apr 2016 17:21:24]] [ERROR]   [RangeError: Maximum call stack size exceeded]
```

I found out the culprit of this issue is the code of the Javascript SDK that traverses through objects looking for unsaved children.

More specifically this piece of code in the `unsavedChildren.js` file.
```
function traverse(
  obj: ParseObject,
  encountered: EncounterMap,
  shouldThrow: boolean,
  allowDeepUnsaved: boolean
) {
  if (obj instanceof ParseObject) {
    if (!obj.id && shouldThrow) {
      throw new Error('Cannot create a pointer to an unsaved Object.');
    }
    var identifier = obj.className + ':' + obj._getId();
    if (!encountered.objects[identifier]) {
      encountered.objects[identifier] = (
        obj.dirty() ? obj : true
      );
      var attributes = obj.attributes;
      for (var attr in attributes) {
        if (typeof attributes[attr] === 'object') {
          traverse(attributes[attr], encountered, !allowDeepUnsaved, allowDeepUnsaved);
        }
      }
    }
    return;
  }
  if (obj instanceof ParseFile) {
    if (!obj.url() && encountered.files.indexOf(obj) < 0) {
      encountered.files.push(obj);
    }
    return;
  }
  if (obj instanceof ParseRelation) {
    return;
  }
  if (Array.isArray(obj)) {
    obj.forEach((el) => {
      traverse(el, encountered, shouldThrow, allowDeepUnsaved);
    });
  }
  for (var k in obj) {
    if (typeof obj[k] === 'object') {
      traverse(obj[k], encountered, shouldThrow, allowDeepUnsaved);
    }
  }
}
```
**What led to the issue:**

The `unsavedChildren` function calls traverse for each attribute on the class. When it's called for an array attribute traverse will be called recursively for each of the variables it contains.
Eventually each string in the array will be processed individually by this piece of code:

```
 for (var k in obj) {
    if (typeof obj[k] === 'object') {
      traverse(obj[k], encountered, shouldThrow, allowDeepUnsaved);
    }
  }
```

For-looping through a string this way will loop through all its attributes, including the styling options such as `strikethrough`, `underline`,`bold`,`blue`,`red` etc. which are all styled versions of the original string itself. 
Since strings are objects the traverse method will traverse these styled strings as well, and it will then find the same styled string attributes, which it will traverse through etc. which will lead to an infinite loop.

**Comment by andrewimm** [comment]
Are your strings actually strings? If this were an issue with strings, these tests would have failed: https://github.com/ParsePlatform/Parse-SDK-JS/blob/master/integration/test/ArrayOperationsTest.js

You mention attributes like `strikethrough`, `underline`, etc. These aren't attributes of `String.prototype`, as defined by ECMAScript specs, which is why I'm curious what objects you're actually saving to your array.

**Comment by andrewimm** [comment]
FWIW, I have found a related issue in `traverse`, though it's not what you describe. The method assumes (rightly) that it is only called on object types, but there is one case where its possible for non-object types to be passed to it. I've patched this up, and assuming tests pass it'll probably solve your problems.

I'd still be curious to know why your strings have all of these other properties, though.

**Comment by Gameleon12** [comment]
Using the traverse fix did solve the issue for me.

However, I am unaware why the string contains all those properties. I use string literals (as shown in the example). It seems one of the node modules I use modifies the string prototype then.

**Comment by andrewimm** [comment]
Fix in 1.8.2 onwards
